### PR TITLE
test: add unit tests for impersonationRequests model

### PR DESCRIPTION
**Pull request by Suvidh-kaushik**
Date: 27/05/2025

Developer Name: Suvidh Kaushik

---

## Issue Ticket Number

Closes: #2200 

## Description
- This is the 3rd PR for feature impersonation
- It contains the tests for impersonationRequests Model, and the fixtures required for it

**This PR will be merged after the PR #2343 has been merged**
**The build will fail since the feature PR is not merged yet**

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [ ] Yes
- [x] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots
<details>
<summary>Screenshot 1</summary>

N/A

</details>

## Test Coverage
<details>
<summary>Screenshot 1</summary>

Screenshot 2025-06-05 003146
Screenshot 2025-06-05 003208

</details>

## Additional Details

Design Doc - [LINK](https://docs.google.com/document/d/1Qt0-knoFEWDl6Gdxr0W1xhPy87H-ixwj9Y55Q0bKMUo/edit?tab=t.0#heading=h.w1nbw3goorza)

PRD - [LINK](https://docs.google.com/document/d/1KI7zp81vevaIX3nDNgTsCHDY1iWmad-XFaVgEi3nH2Q/edit?tab=t.0)

### CODE REVIEW

File: test/unit/models/impersonationRequests.test.ts [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+    });
+  });
+
+  describe("getImpersonationRequests", () => {
+    it("should return a list of impersonation requests", async () => {
+      await createImpersonationRequest(ImpersonationRequest1);
+      await createImpersonationRequest(ImpersonationRequest2);
+      const requests = await getImpersonationRequests({});
+      expect(requests).to.not.be.null;
+      expect(requests.allRequests.length).to.be.greaterThan(0);
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:157]: please check exact number here

File: test/fixtures/impersonation-requests/impersonationRequests.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+export const ImpersonationRequest2={
+  status: REQUEST_STATE.PENDING,
+  isImpersonationAttempted: false,
+  createdBy: "superuser-2",
+  createdFor: "suvidh-kaushik-2",
+  userId: "userId124",
+  reason: "He asked again",
+  impersonatedUserId: "userId445",
+};
```
Comment (AnujChhikara) [inline test/fixtures/impersonation-requests/impersonationRequests.ts:26]: can we please create one object for this

File: test/unit/models/impersonationRequests.test.ts [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        await createImpersonationRequest(ImpersonationRequest1);
+        await createImpersonationRequest(ImpersonationRequest1);
+        expect.fail("Error while creating request");
+      } catch (error) {
+        expect(error).to.exist;
+      }
+    });
+
+    it("should create multiple requests for different users", async () => {
+      const req2 = await createImpersonationRequest(ImpersonationRequest2);
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:60]: why this variable name?

File: test/unit/models/impersonationRequests.test.ts [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+      try {
+        await createImpersonationRequest(ImpersonationRequest1);
+        await createImpersonationRequest(ImpersonationRequest1);
+        expect.fail("Error while creating request");
+      } catch (error) {
+        expect(error).to.exist;
+      }
+    });
+
+    it("should create multiple requests for different users", async () => {
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:59]: what is the purpose of this test like what edge case we are checking?

File: test/unit/models/impersonationRequests.test.ts [hunk 5]
```diff
@@ -0,0 +1,285 @@
...
+          ...wrongCreateImpersonationRequestModelBody,
+        });
+        expect.fail("Should throw error for missing fields");
+      } catch (error) {
+        expect(error).to.exist;
+      }
+    });
+  });
+
+  describe("getImpersonationRequestById", () => {
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:78]: is it possible to add test for throwing error edge case?

File: test/unit/models/impersonationRequests.test.ts [hunk 6]
```diff
@@ -0,0 +1,285 @@
...
+    it("should return null if the request does not exist", async () => {
+      const request = await getImpersonationRequestById({ id: "nonexistentId" });
+      expect(request).to.be.null;
+    });
+  });
+
+  describe("updateImpersonationRequest", () => {
+    beforeEach(async () => {
+      impersonationRequest = await createImpersonationRequest(ImpersonationRequest1);
+    });
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:95]: why are we not having afterEach block here?

File: test/unit/models/impersonationRequests.test.ts [hunk 7]
```diff
@@ -0,0 +1,285 @@
...
+        startedAt: Timestamp.fromDate(new Date()),
+        endedAt: Timestamp.fromDate(new Date(Date.now() + 1000)),
+      };
+      const result = await updateImpersonationRequest(
+        impersonationRequest.id,
+        updatedBody,
+        impersonationRequest.userId
+      );
+      expect(result.isImpersonationAttempted).to.be.true;
+      expect(Number(result.startedAt)).to.be.greaterThan(0);
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:136]: why we are doing this ?

File: test/unit/models/impersonationRequests.test.ts [hunk 8]
```diff
@@ -0,0 +1,285 @@
...
+      await createImpersonationRequest(ImpersonationRequest1);
+      await createImpersonationRequest(ImpersonationRequest2);
+      const requests = await getImpersonationRequests({});
+      expect(requests).to.not.be.null;
+      expect(requests.allRequests.length).to.be.greaterThan(0);
+    });
+
+    it("Should return a list of all the requests with specified status - APPROVED", async () => {
+      const impersonationRequest = await createImpersonationRequest(ImpersonationRequest1);
+      await updateImpersonationRequest(
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:162]: suggestion : - we can do this in before all block where we can create request and approve and reject them and here we can just test them?

File: test/unit/models/impersonationRequests.test.ts [hunk 9]
```diff
@@ -0,0 +1,285 @@
...
+      expect(impersonationRequestData.allRequests[0].status).to.be.equal(REQUEST_STATE.APPROVED);
+    });
+
+    it("Should return a list of all the requests with specified status - PENDING", async () => {
+      await createImpersonationRequest(ImpersonationRequest1);
+      const query = { status: REQUEST_STATE.PENDING };
+      const impersonationRequestData = await getImpersonationRequests(query);
+      expect(impersonationRequestData.allRequests[0].status).to.be.equal(REQUEST_STATE.PENDING);
+    });
+
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:178]: rejected status test missing

File: test/unit/models/impersonationRequests.test.ts [hunk 10]
```diff
@@ -0,0 +1,285 @@
...
+    it("Should return a list of all the requests with specified status - APPROVED", async () => {
+      const impersonationRequest = await createImpersonationRequest(ImpersonationRequest1);
+      await updateImpersonationRequest(
+        impersonationRequest.id,
+        updateImpersonationRequestApproved,
+        impersonationRequest.impersonatedUserId
+      );
+      const query = { status: REQUEST_STATE.APPROVED };
+      const impersonationRequestData = await getImpersonationRequests(query);
+      expect(impersonationRequestData.allRequests[0].status).to.be.equal(REQUEST_STATE.APPROVED);
```
Comment (shubhdevelop) [inline test/unit/models/impersonationRequests.test.ts:169]: have at least couple of entries in the db so that we can assert all the entries in the array have the same status be it "APPROVED", "REJECTED", "PENDING"

File: test/unit/models/impersonationRequests.test.ts [hunk 11]
```diff
@@ -0,0 +1,285 @@
...
+        impersonationRequest.impersonatedUserId
+      );
+      expect(Number(updatedRequest.updatedAt)).to.be.greaterThan(before);
+    });
+  });
+
+  describe("getImpersonationRequests", () => {
+    it("should return a list of impersonation requests", async () => {
+      await createImpersonationRequest(ImpersonationRequest1);
+      await createImpersonationRequest(ImpersonationRequest2);
```
Comment (shubhdevelop) [inline test/unit/models/impersonationRequests.test.ts:154]: as we will require to create multiple requests for testing purpose, take this out in a beforeEach block

File: test/unit/models/impersonationRequests.test.ts [hunk 12]
```diff
@@ -0,0 +1,285 @@
...
+    });
+
+    it("Should return a list of all the requests by size ", async () => {
+      await createImpersonationRequest(ImpersonationRequest1);
+      await createImpersonationRequest(ImpersonationRequest2);
+      const query = { size: 1 };
+      const impersonationRequestData = await getImpersonationRequests(query);
+      expect(impersonationRequestData.allRequests).to.have.lengthOf(1);
+    });
+
```
Comment (shubhdevelop) [inline test/unit/models/impersonationRequests.test.ts:234]: add a case where both the size and page query is being passed

File: test/fixtures/impersonation-requests/impersonationRequests.ts [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+
+export const wrongCreateImpersonationRequestModelBody = {
+  status: REQUEST_STATE.PENDING,
+  isImpersonationAttempted: false,
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/fixtures/impersonation-requests/impersonationRequests.ts:30]: yes will be changed once the types PR get's merged to staging and I rebase

File: test/unit/models/impersonationRequests.test.ts (hunk 3 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:60]: have updated this to request

File: test/fixtures/impersonation-requests/impersonationRequests.ts (hunk 2 above)
Comment (Suvidh-kaushik) [reply test/fixtures/impersonation-requests/impersonationRequests.ts:26]: okay will create an array for this.

File: test/unit/models/impersonationRequests.test.ts (hunk 6 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:95]: added itc

File: test/unit/models/impersonationRequests.test.ts (hunk 11 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:154]: done

File: test/unit/models/impersonationRequests.test.ts (hunk 8 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:162]: okay

File: test/unit/models/impersonationRequests.test.ts (hunk 10 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:169]: okay

File: test/unit/models/impersonationRequests.test.ts (hunk 9 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:178]: added it

File: test/unit/models/impersonationRequests.test.ts (hunk 5 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:95]: okay

File: test/unit/models/impersonationRequests.test.ts (hunk 4 above)
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:65]: okay

File: test/unit/models/impersonationRequests.test.ts [hunk 14]
```diff
@@ -0,0 +1,351 @@
...
+    it("should log and throw error if creation fails", async () => {
+      sinon.stub(impersonationModelService, "createImpersonationRequest");
+      try {
+        await impersonationModelService.createImpersonationRequest(impersonationRequestsBodyData[0]);
+      } catch (err) {
+        expect(err).to.equal(ERROR_WHILE_CREATING_REQUEST);
+        expect(logger.calledWith(ERROR_WHILE_CREATING_REQUEST, ERROR_WHILE_CREATING_REQUEST)).to.be.true;
+      }
```
[the remark this answers is not in the document]
Comment (Suvidh-kaushik) [reply test/unit/models/impersonationRequests.test.ts:94]: i am stubbing the model function so do you think it is necessary to stub the logger also since logging will be a default behaviour when we stub the function

File: test/unit/models/impersonationRequests.test.ts [hunk 15]
```diff
@@ -0,0 +1,343 @@
...
+      } catch (err) {
+        expect(loggerStub.calledWith(ERROR_WHILE_UPDATING_REQUEST, err)).to.be.true;
+      }
+    });
+  });
+
+  describe("getImpersonationRequests", () => {
+    beforeEach(async () => {
+      impersonationRequests = [];
+      for (let i = 0; i < 5; i++) {
```
Comment (AnujChhikara) [inline test/unit/models/impersonationRequests.test.ts:192]: can we use `promise.all` here?

File: test/unit/models/impersonationRequests.test.ts [hunk 16]
```diff
@@ -0,0 +1,346 @@
...
+  impersonationRequestsBodyData,
+  updateImpersonationRequestApproved,
+  updateImpersonationRequestRejected,
+} from "../../fixtures/impersonation-requests/impersonationRequests";
+import { REQUEST_STATE, ERROR_WHILE_CREATING_REQUEST, ERROR_WHILE_FETCHING_REQUEST, ERROR_WHILE_UPDATING_REQUEST } from "../../../constants/requests";
+import { Timestamp } from "firebase-admin/firestore";
+import addUser from "../../utils/addUser";
+import userDataFixture from "./../../fixtures/user/user";
+import sinon from "sinon";
+import { CreateImpersonationRequestModelDto, UpdateImpersonationRequestStatusBody } from "../../../types/impersonationRequest";
```
Comment (Suvidh-kaushik) [inline test/unit/models/impersonationRequests.test.ts:14]: This error is because our model funtions PR is not yet merged in develop and it has this CreateImpersonationRequestModelDto type , once it is merged and i rebase the branch we will not get this error
